File: website/src/features/auth/data/repositories/AuthRepositoryImpl.ts
/**
 * Auth Repository Implementation
 * Data layer - Implements IAuthRepository interface
 */

import {
  IAuthRepository,
  LoginCredentials,
  SignupCredentials,
  AuthResult,
} from '../../domain/repositories/IAuthRepository';
import { User } from '../../domain/entities/User';
import { AuthDataSource } from '../datasources/AuthDataSource';
import { UserModel } from '../models/UserModel';

export class AuthRepositoryImpl implements IAuthRepository {
  private dataSource: AuthDataSource;

  constructor() {
    this.dataSource = new AuthDataSource();
  }

  async signIn(credentials: LoginCredentials): Promise<AuthResult> {
    try {
      const result = await this.dataSource.signIn(credentials.email, credentials.password);

      if (!result.success || !result.user) {
        return {
          success: false,
          error: result.error || 'Login failed',
        };
      }

      // Store remember me preference
      if (credentials.rememberMe !== undefined) {
        localStorage.setItem('rememberMe', credentials.rememberMe ? 'true' : 'false');
      }

      const user = UserModel.fromSupabase(result.user);

      return {
        success: true,
        user,
      };
    } catch (error) {
      return {
        success: false,
        error: error instanceof Error ? error.message : 'An unexpected error occurred',
      };
    }
  }

  async signUp(credentials: SignupCredentials): Promise<AuthResult> {
    try {
      const result = await this.dataSource.signUp(
        credentials.email,
        credentials.password,
        credentials.metadata
      );

      if (!result.success || !result.user) {
        return {
          success: false,
          error: result.error || 'Signup failed',
        };
      }

      const user = UserModel.fromSupabase(result.user);

      return {
        success: true,
        user,
      };
    } catch (error) {
      return {
        success: false,
        error: error instanceof Error ? error.message : 'An unexpected error occurred',
      };
    }
  }

  async signOut(): Promise<{ success: boolean; error?: string }> {
    try {
      const result = await this.dataSource.signOut();
      return result;
    } catch (error) {
      return {
        success: false,
        error: error instanceof Error ? error.message : 'Sign out failed',
      };
    }
  }

  async getCurrentUser(): Promise<User | null> {
    try {
      const result = await this.dataSource.getCurrentUser();

      if (!result.success || !result.user) {
        return null;
      }

      return UserModel.fromSupabase(result.user);
    } catch (error) {
      console.error('Get current user error:', error);
      return null;
    }
  }

  async isAuthenticated(): Promise<boolean> {
    try {
      const sessionResult = await this.dataSource.getCurrentSession();
      return sessionResult.success && sessionResult.session !== null;
    } catch (error) {
      return false;
    }
  }

  async resetPassword(email: string): Promise<{ success: boolean; error?: string }> {
    try {
      const result = await this.dataSource.resetPassword(email);
      return result;
    } catch (error) {
      return {
        success: false,
        error: error instanceof Error ? error.message : 'Password reset failed',
      };
    }
  }
}

File: myFinance_improved_V2/lib/features/auth/data/repositories/auth_repository_impl.dart
// lib/features/auth/data/repositories/auth_repository_impl.dart

import '../../domain/entities/user_entity.dart';
import '../../domain/repositories/auth_repository.dart';
import '../datasources/supabase_auth_datasource.dart';
import '../models/freezed/user_dto_mapper.dart';
import 'base_repository.dart';

/// Auth Repository Implementation
///
/// 📜 계약 이행자 - Domain Repository Interface를 구현
///
/// 책임:
/// - Domain 계약 준수 (AuthRepository interface)
/// - AuthDataSource 호출
/// - Model ↔ Entity 변환
/// - Exception 처리 및 변환 (BaseRepository 상속)
///
/// 이 계층은 Domain과 Data 사이의 변환을 담당합니다.
/// Supabase에 대한 지식은 없으며, AuthDataSource를 통해서만 데이터에 접근합니다.
class AuthRepositoryImpl extends BaseRepository implements AuthRepository {
  final AuthDataSource _dataSource;

  AuthRepositoryImpl(this._dataSource);

  @override
  Future<User?> login({
    required String email,
    required String password,
  }) {
    return executeNullable(() async {
      // Call DataSource
      final userModel = await _dataSource.signIn(
        email: email,
        password: password,
      );

      // Convert Model to Entity
      return userModel.toEntity();
    });
  }

  @override
  Future<User> signUp({
    required String email,
    required String password,
    String? firstName,
    String? lastName,
  }) {
    return execute(() async {
      // Call DataSource
      final userModel = await _dataSource.signUp(
        email: email,
        password: password,
        firstName: firstName,
        lastName: lastName,
      );

      // Convert Model to Entity
      return userModel.toEntity();
    });
  }

  @override
  Future<void> logout() {
    return execute(() => _dataSource.signOut());
  }
}
